### ISSUE TITLE
Suggestion - exclude intermediate selectors

### ISSUE THREAD

**Issue by stylinchas** [less/less.js#2551]
I want to suggest an idea for an enhancement. I've searched around and don't see it so apologies if I missed something like this, and please let me know if I can already do this.

Nested Less greatly improves readability, as it mimics the DOM structure, lets you easily see which styles are being inherited, etc. 

The problem is you end up with overly specific selectors. This:

    .main-nav {
        ul {
            li {
                a {

                }
            }
        }
    }

outputs as:

    .main-nav { } 
    .main-nav ul { } 
    .main-nav ul li { } 
    .main-nav ul li a { } 

It would be great if you could tag selectors to indicate that they get added to the output only if they are the key (last one in the selector sequence). Here I use * to indicate selectors that are only output if they are the key.

    .main-nav {
        ul* {
            li* {
                a {
                }
            }
        }
    }

This would output: 

    .main-nav { } 
    .main-nav ul { } 
    .main-nav li { } 
    .main-nav a { } 

This capability would help keep rule specificity low and make nesting much more useable.

Thanks for considering this idea,

Charlie

Charles Wyke-Smith - charles@stylinwithcss.com

**Comment by seven-phases-max** [comment]
See #2365

**Comment by seven-phases-max** [comment]
Closing as duplicate #2365. (Feel free to reopen if you disagree).

**Comment by matthew-dean** [comment]
I would say this is similar to #1075, in which parent selectors can be selected (which is where that issue ended up leaning), or, perhaps the ```&``` statement could be altered (which sounds easier than it actually would be, so it's unlikely).

All of these issues have a core problem, which is that people want to structure Less "organizationally", and have that be somewhat different to how those structures are collapsed into CSS structures.

To be honest, it's not an easy problem. For your example, the thing that jumps out at me is that ```ul*``` is far too similar to ```ul *``` which would be a valid selector.

I wonder if, rather than parent targets, which can be ambiguous (and was noted as an issue in #1075), we could do something like "capture groups" (similar to regex, but just to mark a section), as in.

```
{.main-nav} {
    ul {
        li {
            &{1} a {
             
            }
        }
    }
}
```
We had discussed numerical referencing of ```&```, but because of its ambiguity in some cases, it was dropped. If, however, you could very clearly "mark" a selector that you want to select later to append to, that might make sense to target specifically rather than trying to exclude specific selectors one by one. Just an idea.

*Note to @seven-phases-max: this doesn't look like a duplicate of #2365, because the question isn't about automatic removal of selectors.*

**Comment by seven-phases-max** [comment]
I don't think it's #1075. Aithour's code is just:
```
.main-nav {
   ul {}
   li {}
   a {}
}
```
But he wants redundant nesting because he believes that:
>Nested Less greatly improves readability, as it mimics the DOM structure

(w/o arguing that particular style of coding) I suppose in that sense it's pretty much the same as in #2365 - regardless if it's automatic or explicitly set with some special symbol.

**Comment by matthew-dean** [comment]
Well, while he's collapsing the tree, essentially what he's saying is he wants some way to reference / import / append only ```.main-nav``` to append to ```a```. So, he's not directly asking for a parent selector, but it's essentially the same problem. i.e. wanting to structure a tree differently without producing the same output. We've seen variations of this with different requests and different suggested solutions, such as parent targets or automatic collapsing or extend keywords, or, in this case, tagged exclusion, but they're really all one problem.

It's natural to author Less this way, because it allows nesting, and so does HTML. So you can do a one to one "map" to modular code. But of course, many Less and Sass tutorials argue against it, because it produces unnecessary selectors. In fact, that's the rationale some give for avoiding Less or Sass or Stylus altogether: because you will naturally create this structure, and this structure produces verbose output.

It's hard to blame coding style when it's intuitive to do this. I haven't seen a Less project that didn't nest in order to represent the DOM structure (instead of producing/targeting just a particular desired CSS output).

So, that's a bit of a larger problem than this question; it's just to say that this question is germane to that problem.

**Comment by matthew-dean** [comment]
We could "collapse" (close) this one against the other issues that best reference the overall problem, OR we could create a new issue that encompasses the scope of the others and close those? What do you think?

**Comment by seven-phases-max** [comment]
So what about "closing as duplicate of #1075 *or* #2365 (whatever you find suitable)"?

**Comment by matthew-dean** [comment]
They're both referenced now from this issue, so regardless, let's close as likely duplicate of one of them :smile: .

**Comment by stylinchas** [comment]
Matthew,
Thanks for the detailed reply. I appreciate you considering this suggestion.

As for my “exclude selector if not key” symbol, I just used * after the selector element as a suggestion - any unambiguous character before or after could work - so don’t let that put you off thinking about doing this!

Also, I don’t, as Max suggests, "propose you intentionally use wrong nesting when you never need it at all and then want a special mark to turn the "wrong" into the "right”. “. The nesting would be correct (i.e. match the HTML). I just want to be able to exclude superfluous intermediate tags that would otherwise be generated in the output.

The reason that I think this would be a huge feature is that it’s logical, before starting to add styles, to mimic the nesting of the HTML with the Less structure, using just the selectors and curly braces. If the page is complex, I’ll do this in a number of blocks that match the main structural HTML blocks, rather than as a single huge nested layout.

I can then focus on adding styles in this Less hierarchy wherever I need them, and not be distracted by having to look at the HTML all the time. Being able to do this is for me of one of the biggest timesavers of Less over CSS. It’s also easy to see what styles are flowing down from the ancestor elements.

The downside, of course, is that you end up with unnecessarily long, and presumably slower, selectors for many of the style rules. Most of the intermediate ones can be removed.

Being able to exclude those unneeded intermediate selectors from the output would very helpful. Specificity management (it’s a thing :) is a key CSS technique. I always want the lowest specificity possible - ideally, only comprising of the class of the containing structural element, and the targeted element itself. Low specificity makes it easy to compare similar selectors and see the winner, and most importantly, helps prevent one rule accidentally crushing another. So I always imagine myself adding those little symbols and having the unwanted selectors disappear.

Hope this helps you see why I would like to see this feature - I’ve got to think I’m not the only one.

Cheers,

Charlie

And yes, this is different from issue #2365 because that process is automatically removing unneeded IDs whereas I want to be able chose the elements in the selector that I want to eliminate.

**Comment by seven-phases-max** [comment]
(Clearly an offtopic here but just a remark for future visitors)
@stylinchas  I hope you realize that with this trick you only mechanically remove redundant parts of the generated selectors but the bloatness and over-specificity of your *source -code* remains. Sadly it's usually pointless to try to make somebody to change his mind about the coding style he once chose (sure it's more "readable" for you if you used it for years) but it's really really arguable that
```
.main-nav {
    ... {
        ... {
            a {
            }
        }
    }
}
``` 
in anyway is more readable and/or structured than even plain CSS
```
.main-nav a {}
```
(and the nesting is normally only a method to avoid `.main-nav` repetition).
This trick is especially becomes weird since as soon as it generates`.main-nav a` you would never guess that styles that may also apply to `.main-nav div div div a` element are buried somewhere under `.main-nav { ul { li { div { span { i { a` spagetti.
Well, this "Mimicing HTML in CSS" anti-pattern clearly requires a whole dedicated blog-post/article so it's just too much of details and pros/cons dismantling one can write in a single issue tracker post... so maybe I'll just leave a link to something to start reading with [Nested selectors: the inception rule](http://thesassway.com/beginner/the-inception-rule#css-selector-nightmare) and you'll find more related articles yourself if interested.

**Comment by lukeapage** [comment]
What about this already supported syntax (if i understand?) 

```
.main-nav {
    ul {
   } 
    & /*ul*/ {
        li {
            a {
            }
        }
    }
}
```
A bit more verbose i know...

**Comment by matthew-dean** [comment]
@lukeapage - Yeah, that occurred to me too. You could do that, or even just note the HTML structure of the module above this block (as a comment). You could also do something like:
```less
.main-nav {
    // ul {
        // li {
            a {

            }
       //  }
    // }
}
```
Of course, that wouldn't work if ul and li had props / vals, but then neither would the "exclusion" model. FYI, this is what I meant by a documented dom model
```less
/// .main-nav ul li a
.main-nav {
     a {

     }   
}
```
So, there's ways to illustrate HTML structure without adding a Less feature in this instance. Maybe, as @seven-phases-max says, it's different enough that it doesn't have merit enough on its own since there are many ways to illustrate the same thing.
